bin/hdfcoinc/pycbc_multiifo_combine_statmap

```diff
@@ -24,23 +24,58 @@ files = [h5py.File(n, 'r') for n in args.statmap_files]
 f = h5py.File(args.output_file, "w")
 
 logging.info('Copying segments and attributes to %s' % args.output_file)
-# Move segments information into the final file - remove some duplication in earlier files
+# Move segments information into the final file - remove some duplication
+# in earlier files
 for fi in files:
     for key in fi['segments']:
         if key.startswith('foreground') or key.startswith('background'):
             continue
         f['segments/%s/end' % key] = fi['segments/%s/end' % key][:]
         f['segments/%s/start' % key] = fi['segments/%s/start' % key][:]
         if 'segments/foreground_veto' in fi:
-            f['segments/%s/foreground_veto/end' % key] = fi['segments/foreground_veto/end'][:]
-            f['segments/%s/foreground_veto/start' % key] = fi['segments/foreground_veto/start'][:]
+            f['segments/%s/foreground_veto/end' % key] = \
+                                         fi['segments/foreground_veto/end'][:]
+            f['segments/%s/foreground_veto/start' % key] = \
+                                       fi['segments/foreground_veto/start'][:]
         for attr_name in fi.attrs:
             if key not in f:
                  f.create_group(key)
             f[key].attrs[attr_name] = fi.attrs[attr_name]
 
+logging.info('combining foreground and foreground excluded segments')
+# Set up dictionaries to contain segments from the individual statmap files
+indiv_segs = segments.segmentlistdict({})
+indiv_segs_fg_exc = segments.segmentlistdict({})
+
+# obtain list of all ifos involved in the coinc_statmap files
+all_ifos = numpy.unique([ifo for fi in files for ifo in \
+                         fi.attrs['ifos'].split(' ')])
+
+# loop through statmap files and put segments into segmentlistdicts
+for fi in files:
+    key = fi.attrs['ifos'].replace(' ','')
+    # get analysed segments from individual statmap files
+    starts = fi['segments/{}/start'.format(key)][:]
+    ends = fi['segments/{}/end'.format(key)][:]
+    indiv_segs[key] = pycbc.events.veto.start_end_to_segments(starts, ends)
+    # get the foreground_vetoed times from individual statmap files
+    remove_start_time = fi['segments/foreground_veto/start'][:]
+    remove_end_time = fi['segments/foreground_veto/end'][:]
+    vetoed_segs = pycbc.events.veto.start_end_to_segments(remove_start_time,
+                                                          remove_end_time)
+    # remove from analysis segments
+    indiv_segs_fg_exc[key] = indiv_segs[key] - vetoed_segs
+
+# Add the segments to obtain '2 or more detectors on' time
+foreground_segs = numpy.sum(list(indiv_segs.values()), axis=0)
```

Comment:
can you explain a bit how this addition works in terms of the structures (dicts, lists, segments, ...)

Reply:
Is that a request for better inline comments or for detailed explanation here?

`indiv_segs` is a segmentlistdict, a dictionary containing the segmentlists from the individual statmap files, in this case it is keyed on detector combination

Using `list(indiv_segs.values())` we get a list of segmentlists

We then add these together using `numpy.sum` and by setting `axis=0` we end up getting `segmentlist[0] + segmentlist[1] + ...` rather than `sum(segmentlist[0]) + sum(segmentlist[1]) + ...` as is the `numpy.sum` default, and doesn't really make sense for segmentlists. As it is a segmentlist type, this addition then works in the natural segment addition way

`indiv_segs_fg_exc` is the same but with the foreground excluded. By adding these segmentlists together we get the foreground time, but with times removed which have triggers in *all currently 'on'* detectors. So a 2-detector coincidence in 3-detector time would *not* be removed unless it has an associated 3-detector coincidence. There is a question which can be asked here about the window size given the difference in how the different foreground exclusive windows are formed in the `coinc_statmap` stage but that may not be a necessary conversation

These segments then get turned into a total time by the `abs` function in the output lines

I hope this is clear, if not then ask

Reply:
This sounds like a long enough explanation that some comments would be good, eg 

```
# Convert segmentlistdict to a list ('seglists') of segmentlists
# then numpy.sum(seglists, axis=0) does seglists[0] + seglists[1] + ...
# Note that 'exclusive' time only removes windows around coincs
# that involve *all* active ifos (2 in 2-ifo time, 3 in 3-ifo time ..)
```

Reply:
comments added in